### PR TITLE
feat: segments cleanup

### PR DESCRIPTION
**Pull request by aliszka**
### What's being changed:
Introduces segment cleaner, its goal is to remove from segments keys that were updated or deleted in newer segments. 
Usually obsolete keys are deleted while 2 segments are compacted, so merged segment does not contain keys updated/tombstoned in right segment. Cleaner verifies existence of keys not only in closest segment, but in all newer segments created after processed segment.
It could be used in environments where compaction is no longer possible (like maximum size of single segment is reached and can not be compacted further) providing other way of keeping segments compact and clean of outdated redundant data.

Cleanup method is started by the same routine that starts compaction process, being secondary to compaction.
If compaction is not performed within compaction cycle (there were no candidate segments to compact), cleanup will be attempted. If compaction is performed, cleanup will be skipped.
Cleanup cleans just one segment per cycle, moving to next one in the following cycle. Names of segments being cleaned, time of cleaning and size of cleaned segment are saved to `cleanup.db.bolt` file, stored in relevant bucket directory.

Cleanup should be enabled by setting interval between another clean attempts. Interval is set by env variable `PERSISTENCE_LSM_SEGMENTS_CLEANUP_INTERVAL_HOURS`. By default variable is set to `0`, effectively having cleanup turned off. 

Same PR to stable/v1.26: https://github.com/weaviate/weaviate/pull/5953

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/handlers/rest/configure_api.go [hunk 1]
```diff
@@ -289,26 +289,27 @@ func MakeAppState(ctx context.Context, options *swag.CommandLineOptionsGroup) *s
...
-		DisableLazyLoadShards:     appState.ServerConfig.Config.DisableLazyLoadShards,
-		ForceFullReplicasSearch:   appState.ServerConfig.Config.ForceFullReplicasSearch,
+		ServerVersion:                config.ServerVersion,
+		GitHash:                      config.GitHash,
+		MemtablesFlushDirtyAfter:     appState.ServerConfig.Config.Persistence.MemtablesFlushDirtyAfter,
+		MemtablesInitialSizeMB:       10,
+		MemtablesMaxSizeMB:           appState.ServerConfig.Config.Persistence.MemtablesMaxSizeMB,
+		MemtablesMinActiveSeconds:    appState.ServerConfig.Config.Persistence.MemtablesMinActiveDurationSeconds,
+		MemtablesMaxActiveSeconds:    appState.ServerConfig.Config.Persistence.MemtablesMaxActiveDurationSeconds,
+		SegmentsCleanupIntervalHours: appState.ServerConfig.Config.Persistence.LSMSegmentsCleanupIntervalHours,
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/handlers/rest/configure_api.go:299]: Will change it to seconds to keep consistency. Env variable though would still be accepting hours though

File: adapters/repos/db/lsmkv/cleanup_replace_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,813 @@
...
+
+	type kvt struct {
+		pkey string
+		val  string
+		tomb bool
+	}
+
+	t.Run("create segments", func(t *testing.T) {
+		/*
+			SEG1	SEG2	SEG3	SEG4	SEG5
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/lsmkv/cleanup_replace_integration_test.go:44]: thx

File: adapters/repos/db/lsmkv/segment_group.go [hunk 3]
```diff
@@ -537,3 +553,38 @@ func fileExists(path string) (bool, error) {
 
 	return false, err
 }
+
+func (sg *SegmentGroup) compactOrCleanup(shouldAbort cyclemanager.ShouldAbortCallback) bool {
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/lsmkv/segment_group.go:575]: I did not, but tests on big dataset have shown no issues. Indeed more segments need to be checked comparing to compaction process, but that is only verifying key existence using embedded tree index without actual reading node's data.

File: adapters/repos/db/lsmkv/segment_cleaner_replace.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	bufw                *bufio.Writer
+	cursor              *segmentCursorReplace
+	keyExistsFn         keyExistsOnUpperSegmentsFunc
+	version             uint16
+	level               uint16
+	secondaryIndexCount uint16
+	scratchSpacePath    string
+}
+
+func newSegmentCleanerReplace(w io.WriteSeeker, cursor *segmentCursorReplace,
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/lsmkv/segment_cleaner_replace.go:36]: that is correct. values of matching keys in newer segments overwrite values stored in older segments.
(that is valid for replace strategy, where values are unique)

File: adapters/repos/db/lsmkv/segment_group_cleanup.go [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+
+	candidateIdx, startIdx, onCompleted, err := c.findCandidate()
+	if err != nil {
+		return false, err
+	}
+	if candidateIdx == noIdx {
+		return false, nil
+	}
+
+	if c.sg.allocChecker != nil {
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/lsmkv/segment_group_cleanup.go:360]: no, exactly same checks are utilized in cleanup process as in compaction

File: adapters/repos/db/lsmkv/segment_group_cleanup.go [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("deleting from cleanup bolt db %q: %w", c.db.Path(), err)
+		}
+	}
+	return nil
+}
+
+func (c *segmentCleanerCommon) readEarliestCleaned(ids, sizes []int64, nowTs int64,
```
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/lsmkv/segment_group_cleanup.go:231]: I will refactor the method and add comment describing all selection rules

File: usecases/config/environment.go [hunk 7]
```diff
@@ -195,6 +195,14 @@ func FromEnv(config *Config) error {
 		config.Persistence.LSMMaxSegmentSize = DefaultPersistenceLSMMaxSegmentSize
 	}
 
+	if err := parsePositiveInt(
+		"PERSISTENCE_LSM_SEGMENTS_CLEANUP_INTERVAL_HOURS",
```
Comment (antas-marcin) [inline usecases/config/environment.go:199]: I think it should be `*_SECONDS` not `*_HOURS` yes?

Comment (aliszka) [reply usecases/config/environment.go:199]: User facing setting is measured in hours for cleanup not to be executed to often. It is internally converted to seconds.

File: adapters/repos/db/lsmkv/segment_group_cleanup.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (aliszka) [reply adapters/repos/db/lsmkv/segment_group_cleanup.go:420]: actually implicitly we do: if available disk space is running low shards go into readonly mode. in that case compaction and now cleanup will not run

**Review by amourao — APPROVED** [review]
LGTM

**Pull request merged**